### ISSUE TITLE
Cannot recover from GPU mismatch exception

### ISSUE THREAD

**Issue by kmaehashi** [chainer/chainer#4144]
The following code raises unexpected `CUDARuntimeError`.

```
import chainer
import chainer.functions as F
import chainer.links as L
import cupy as cp

l = L.Linear(10)
l.to_gpu(0)
arr = cp.arange(10, dtype=cp.float32).reshape((10, 1))
varr = F.copy(arr, 1)

l(varr)

# Raises `ValueError: Array device must be same as the current device: array device = 0 while current = 1` as expected

print(varr)

# Raises `cupy.cuda.runtime.CUDARuntimeError: cudaErrorIllegalAddress: an illegal memory access was encountered`, which is not expected
```

**Comment by kmaehashi** [comment]
This issue should be fixed if feature mentioned in #4459 is implemented (so that it is guaranteed that the current device and all array arguments are the same).

Also raised an CuPy issue in #1389.

**Comment by kmaehashi** [comment]
Merge the discussion to #4459.
